### PR TITLE
fix: rewrite transfer feature files from user perspective, add missing details

### PR DESCRIPTION
**Pull request by cooper-joe**
adjusted feature files to be written from the context, rather than the implications.

also took the opportunity to add some more detail to features that were lacking it.

### CODE REVIEW

**Comment by varl** [comment]
The most important structure to me is the same principle we use in commit messages, that is, don't explain what the commit _did_, explain what it _does_, so instead of explaining what happened:

>When one of the highlighted items is **clicked and dragged**

I prefer that this explains what is happening: 

>When the user **clicks and drags** one of the highlighted items to a new position

Which also makes sense in the test case, where the statement matches what the code _does_, not what it _did_.

```
When('the user clicks and drags one of the highlighted items to a new position', () => {
  // code that clicks and drags a highlighted item from x to y
})
```

If we use "the user" as an actor or not is not that important to me, but the present/past tense form matters a lot for readability and understanding.

**Comment by Mohammer5** [comment]
@varl the following sentence is present:

> When one of the highlighted items is clicked and dragged

It's passive, not past. 

These are virtually identical in meaning:

> When one of the highlighted items is clicked and dragged by the user
> When the user clicks and drags one of the highlighted items

I like having passive sentences without the acting entity as that describes pure behavior regardless of who is acting (User, accessibility software, e2e test engine, etc)

**Comment by varl** [comment]
> @varl the following sentence is present:
> 
> > When one of the highlighted items is clicked and dragged
> 
> It's passive, not past.

OK, I may be using the wrong grammar terms here, that is not the heart of the point however.

> 
> These are virtually identical in meaning:
> 
> > When one of the highlighted items is clicked and dragged by the user
> > When the user clicks and drags one of the highlighted items

Agreed, there are many ways to write a sentence so that the meaning is identical. This is also besides the point.

> I like having passive sentences without the acting entity as that describes pure behavior regardless of who is acting (User, accessibility software, e2e test engine, etc)

Passive sentences for what part of Given-When-Then? Or all of them? A common trick that works pretty well in practice is to use past tense for *Given*, present tense for *When*, and future tense for *Then*.

In addition, *Given* and *Then* should be passive, but *When*, should definitely be active.

```
Given the user is logged in as an admin

When the user clicks on delete user "foo"

Then the foo user should be deleted
```

I do not agree that the *When* clause should be passive nor past tense, but I agree with you on *Given* and *Then*.

**Comment by varl** [comment]
Ref. avoiding naming the actor:

The Actor (Cypress, Selenium, Accessibility software) acts on the system _on behalf_ of someone, and that someone is referred to as a generic Role). In UI, we only have a single role (right now at least) and that is a User.

So by not naming the role in the feature files, we just imply it. If in the future we have additional roles, then we will have to make it explicit anyway. To me it doesn't matter that much if we do it now or later.

What is important is to be aware that in applications that have multiple roles _we have to be explicit_ about what role the actor is acting on behalf of. E.g. in the Maintenance app features we have different behaviors depending on if the actor (Cypress) is acting as a restricted user or a super user. 

The actor is constant, the role is not, and that is why it needs to be explicitly mentioned in the feature files.

The question is if we should allow ourselves to have different ways of writing Given-When-Then statements for libraries and applications or be consistent across the board.

**Comment by Mohammer5** [comment]
## Tense

I definitely agree that it makes sense to have `Given` in the past tense. They describe the current situation, so if they're using an action to get there it has to have happened in the past.

I don't think the `Then` step should be future but rather affirmative present (with `should` just like your example) and passive as we're describing a state rather than an action.

In terms of tense, it's probably the easiest to keep them all in the present (you did that in your example as well):

> Given the user is logged in as an admin

## Passive/active

I think you're right here. The `When` step should be active. I've read some articles about what's the standard practice and it seems everyone else is using active here as well (with some good arguments for it).

Another argument for why it's totally fine to include the term "user" is because the entity interacting with the environment is a user, regardless of what is is, even cypress uses the interface to do something. It should suffice to mention a role or specify a particular type of user when necessary

I think we could add a `howto` or `guide` to our notes repo, so we have something to point to and remember this?

**Comment by cooper-joe** [comment]
> I think we could add a `howto` or `guide` to our notes repo, so we have something to point to and remember this?

+1 for this. It would be very useful to be able to refer to this document while writing features and while discussing them in PRs.

File: cypress/integration/Transfer/disabled-transfer-buttons.feature [hunk 1]
```diff
@@ -0,0 +1,41 @@
+Feature: Disable transfer buttons when actions are not possible
+
+  Scenario: None of the available options is highlighted
+      Given the options list has items
+      When no option items are highlighted
```
Comment (Mohammer5) [inline cypress/integration/Transfer/disabled-transfer-buttons.feature]: Should be `And` (`Given`)

File: cypress/integration/Transfer/disabled-transfer-buttons.feature [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      Then the 'move to selected list' button is disabled
+
+  Scenario: Some of the available options is highlighted
+      Given the options list has items
+      When some option items are highlighted
```
Comment (Mohammer5) [inline cypress/integration/Transfer/disabled-transfer-buttons.feature]: Same as above

File: cypress/integration/Transfer/disabled-transfer-buttons.feature [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      Then the 'move to selected list' button is enabled
+
+  Scenario: None of the selected options is highlighted
+      Given the selected list has items
+      When no selected items are highlighted
```
Comment (Mohammer5) [inline cypress/integration/Transfer/disabled-transfer-buttons.feature]: Same as above

File: cypress/integration/Transfer/disabled-transfer-buttons.feature [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      Then the 'move to options list' button is disabled
+
+  Scenario: Some of the selected options is highlighted
+      Given the selected list has items
+      When some selected items are highlighted
```
Comment (Mohammer5) [inline cypress/integration/Transfer/disabled-transfer-buttons.feature]: Same as above

**Comment by Mohammer5** [comment]
I think it should be fine to merge this, then I can reword the scenarios to match our newly defined rules

**Pull request merged**